### ISSUE TITLE
Add note to Writing plugins doc about extracting out utils

### ISSUE THREAD

**Issue by dryoma** [stylelint/stylelint#1365]
Right now only [some stylelint utils](http://stylelint.io/developer-guide/plugins/#stylelintutils) are accessible from plugins. Is it because utils tend to change often/dramatically thus breaking 3rd party code, or maybe there is some other reason?

Yet it would be convenient to have all this stuf in plugins rules.

**Comment by evilebottnawi** [comment]
@dryoma which utils are not accessible?

**Comment by dryoma** [comment]
Only `cssWordIsVariable`, `report`, `ruleMessages`, `styleSearch`, `validateOptions` [are imported by stylelint](https://github.com/stylelint/stylelint/blob/master/src/index.js#L5), as the docs state. And only those can be used in a plugin code by 

```js
import { utils } from "stylelint"

utils.validateOptions( ... ) // ok
utils.isStandardSelector ( ... ) // isStandardSelector is undefined
```

And I was wondering if [the other methods](https://github.com/stylelint/stylelint/tree/master/src/utils) could be made accessible the same way.

**Comment by evilebottnawi** [comment]
@dryoma hm, maybe you're right, let's wait @davidtheclark and @jeddy3 , now you can copy code 	
needed utils and use, after discussion remove code and write `import`. Good?

**Comment by dryoma** [comment]
Yeah, for now I'll have to stick to that variant.

**Comment by davidtheclark** [comment]
> Is it because utils tend to change often/dramatically thus breaking 3rd party code

Yes, we have been selective about which utils are exposed so that we can better control releases and prevent even more releases, especially with breaking changes. 

I'd say that the way to do this would be to create a separate package for stylelint-utils, which can be used by both stylelint and 3rd-party code and has its own releases. That way regular stylelint users will never have to bother upgrading simply because exposed utils changed. It may add a little annoyance to developing stylelint itself, but may be worth it. What do you @stylelint/core think?

**Comment by jeddy3** [comment]
Perhaps we should continue down the path forged by [`postcss-resolve-nested-selector`](https://github.com/davidtheclark/postcss-resolve-nested-selector) i.e. pull applicable utilities into their own plugins and/or npm modules. e.g.

PostCSS based:

- `postcss-is-standard-syntax-rule`
- `postcss-find-font-family`
- `postcss-block-string`
- `postcss-at-rule-param-index`
- etc...

Non-PostCSS based:

- `style-blur-comments`
- `style-search`
- etc...

Stylelint specific non-PostCSS based:

- `stylelint-whitespace-checker`
- etc...

Utilities which we keep within stylelint *but do not expose*:

- `configurationError`
- `styleTagsFromHtmlExtracter`
- etc...

Utilities which we keep within stylelint *but do expose*:

- `ruleMessages`
- `validateOptions`
- `report`

These latter utils are pretty much required to be used by plugins, and don't make sense outside of stylelint. I think we can keep these remaining few within stylelint itself, rather than pushing them into a `stylelint-utils` repo.

What do you think, is this approach more inline with the PostCSS *way of doing things" than creating a `stylelint-utils` repo?

**Comment by davidtheclark** [comment]
Yeah, this is probably the most robust solution. Lots of overhead, but may be worth it.

**Comment by jeddy3** [comment]
I think we can do this as a gradual process. E.g. add a note to the "Writing plugins" doc that says have a look through the built-in utils and if there's something you want to use then consider helping us extract it out. Once it's extracted out, we add it list on that page.

We might find that the `stylelint-scss` guys cherry-pick a few to start. Whereas something like styleSearch or the whitespace checker might come later down the line.

I think the only thing that needs to be done now to set us on this path is to deprecate the exposure of styleSearch, so its exposure can be removed in `v7`. That should stabilise the `stylelint.utils.*` API.

Sound good?

**Comment by davidtheclark** [comment]
Sure!
